spring-batch-infrastructure/src/main/java/org/springframework/batch/item/json/JsonFileItemWriter.java

```diff
@@ -0,0 +1,132 @@
+ * {@link JsonObjectMarshaller}. Items will be enclosed in a json array as follows:
+ *
+ * <p>
+ * <code>
+ * [
+ *  {json object},
+ *  {json object},
+ *  {json object}
+ * ]
+ * </code>
+ * </p>
+ *
+ * The implementation is <b>not</b> thread-safe.
+ *
+ * @see GsonJsonObjectMarshaller
+ * @see JacksonJsonObjectMarshaller
+ * @param <T> type of object to write as json representation
+ * @author Mahmoud Ben Hassine
+ * @since 4.1
+ */
+public class JsonFileItemWriter<T> extends AbstractFileItemWriter<T> {
+
+	private static final char JSON_OBJECT_SEPARATOR = ',';
+	private static final char JSON_ARRAY_START = '[';
+	private static final char JSON_ARRAY_STOP = ']';
+
+	private JsonObjectMarshaller<T> jsonObjectMarshaller;
+
+	/**
+	 * Create a new {@link JsonFileItemWriter} instance.
+	 * @param resource to write json data to
+	 * @param jsonObjectMarshaller used to marshal object into json representation
+	 */
+	public JsonFileItemWriter(Resource resource, JsonObjectMarshaller<T> jsonObjectMarshaller) {
+		Assert.notNull(resource, "resource must not be null");
+		Assert.notNull(jsonObjectMarshaller, "json object marshaller must not be null");
+		setResource(resource);
+		setJsonObjectMarshaller(jsonObjectMarshaller);
+		setHeaderCallback(writer -> writer.write(JSON_ARRAY_START));
+		setFooterCallback(writer -> writer.write(this.lineSeparator + JSON_ARRAY_STOP + this.lineSeparator));
+		setExecutionContextName(ClassUtils.getShortName(JsonFileItemWriter.class));
+	}
+
+	/**
+	 * Assert that mandatory properties (jsonObjectMarshaller) are set.
+	 *
+	 * @see org.springframework.beans.factory.InitializingBean#afterPropertiesSet()
+	 */
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		if (this.append) {
+			this.shouldDeleteIfExists = false;
+		}
+	}
+
+	/**
+	 * Set the {@link JsonObjectMarshaller} to use to marshal object to json.
+	 * @param jsonObjectMarshaller the marshaller to use
+	 */
+	public void setJsonObjectMarshaller(JsonObjectMarshaller<T> jsonObjectMarshaller) {
+		this.jsonObjectMarshaller = jsonObjectMarshaller;
+	}
+
+	@Override
+	public void write(List<? extends T> items) throws Exception {
+		if (!getOutputState().isInitialized()) {
+			throw new WriterNotOpenException("Writer must be open before it can be written to");
+		}
+
+		if (logger.isDebugEnabled()) {
+			logger.debug("Writing to json file with " + items.size() + " items.");
+		}
+
+		OutputState state = getOutputState();
+
+		StringBuilder lines = new StringBuilder();
+		int lineCount = 0;
+		Iterator<? extends T> iterator = items.iterator();
+		while (iterator.hasNext()) {
+			if (iterator.hasNext() && lineCount == 0 && state.getLinesWritten() > 0) {
+				lines.append(JSON_OBJECT_SEPARATOR).append(this.lineSeparator);
+			}
+			T item = iterator.next();
+			lines.append(' ').append(this.jsonObjectMarshaller.marshal(item));
+			if (iterator.hasNext()) {
+				lines.append(JSON_OBJECT_SEPARATOR).append(this.lineSeparator);
+			}
+			lineCount++;
```

Comment:
Everything in this method is the same between the two implementations other than the code within the loop.  Thoughts on adding a `doWrite` method in the abstract parent that takes a single item and returns a `String` and moving the rest of the logic up to the parent's `write`?

Reply:
I thought about it but the structure of the loop is different. One is using a `for` loop and the other is using an iterator + while (which is required to check the existence of a next item if any). But I managed to move the common code in a `doWrite` method in the parent class (which simplified the implementation by the way, see 814734e). Good catch!